File: sdk.py
# sdk.py - Trace Python SDK with API Key

import os
from typing import Any, Dict

import httpx


DEFAULT_BASE_URL = "http://localhost:8000"
DEFAULT_API_KEY = "ag_live_default1234567890"

class AgentGateError(Exception):
    """Base error for Trace/AgentGate SDK."""
    pass

class AuthenticationError(AgentGateError):
    """Raised when API key is invalid."""
    pass


class Trace:
    def __init__(self, api_key: str, base_url: str = DEFAULT_BASE_URL):
        if not api_key or not api_key.strip():
            raise ValueError("Trace requires a non-empty API key")

        self.base_url = base_url.rstrip("/")
        self.api_key = api_key.strip()
        self.client = httpx.Client(timeout=120.0)

    def __enter__(self) -> "Trace":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

    @classmethod
    def from_env(cls, base_url: str = DEFAULT_BASE_URL) -> "Trace":
        # Prefer new TRACE_* env vars but fall back to AGENTGATE_* for compatibility
        env_base = os.getenv("TRACE_BASE_URL", os.getenv("AGENTGATE_BASE_URL", base_url))
        api_key = os.getenv("TRACE_API_KEY", os.getenv("AGENTGATE_API_KEY", DEFAULT_API_KEY))
        return cls(
            base_url=env_base,
            api_key=api_key,
        )

    def _headers(self) -> Dict[str, str]:
        return {
            "Content-Type": "application/json",
            "X-API-Key": self.api_key,
        }

    @staticmethod
    def _raise_for_error(response: httpx.Response) -> None:
        if response.status_code == 401:
            raise AuthenticationError("Authentication failed: Invalid or missing API key")
        if response.status_code != 200:
            try:
                error_detail = response.json()
            except ValueError:
                error_detail = response.text
            raise AgentGateError(f"AgentGate error {response.status_code}: {error_detail}")

    @staticmethod
    def _payload(
        tool: str,
        params: Dict[str, Any],
        agent_id: str,
        policy: Dict[str, Any] | None = None,
        execution_state: Dict[str, Any] | None = None,
        action: str | None = None,
        failure_policy: Dict[str, Any] | None = None,
        entity_resolution: Dict[str, Any] | None = None,
        context: Dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Dict[str, Any]:
        payload = {
            "tool": tool,
            "params": params,
            "agent_id": agent_id,
        }
        if policy:
            payload["policy"] = policy
        if execution_state is not None:
            payload["execution_state"] = execution_state
        if action is not None:
            payload["action"] = action
        if failure_policy is not None:
            payload["failure_policy"] = failure_policy
        if entity_resolution is not None:
            payload["entity_resolution"] = entity_resolution
        if context is not None:
            payload["context"] = context
        if idempotency_key is not None:
            payload["idempotency_key"] = idempotency_key
        return payload

    def call(
        self,
        tool: str,
        params: Dict[str, Any],
        agent_id: str = "default_agent",
        policy: Dict[str, Any] | None = None,
        execution_state: Dict[str, Any] | None = None,
        action: str | None = None,
        failure_policy: Dict[str, Any] | None = None,
        entity_resolution: Dict[str, Any] | None = None,
        context: Dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Dict[str, Any]:
        """Call a tool through the gateway."""
        try:
            response = self.client.post(
                f"{self.base_url}/call",
                json=self._payload(
                    tool,
                    params,
                    agent_id,
                    policy,
                    execution_state,
                    action,
                    failure_policy,
                    entity_resolution,
                    context,
                    idempotency_key,
                ),
                headers=self._headers(),
            )
        except (httpx.ConnectError, httpx.ReadError, httpx.RemoteProtocolError) as exc:
            raise AgentGateError(f"Network error connecting to AgentGate at {self.base_url}: {exc}")
        self._raise_for_error(response)

        return response.json()

    def list_tools(self) -> Dict[str, Any]:
        """List registered tools and capability cards."""
        response = self.client.get(f"{self.base_url}/tools", headers=self._headers())
        self._raise_for_error(response)
        return response.json()

    def to_openai_tools(self, query: str | None = None) -> list[dict[str, Any]]:
        """Format AgentGate tools for OpenAI's 'tools' parameter."""
        tools = self.discover(query)["tools"] if query else self.list_tools()["tools"]
        return [
            {
                "type": "function",
                "function": {
                    # OpenAI names must match ^[a-zA-Z0-9_-]+$
                    "name": t["id"].replace(".", "__").replace("/", "---"),
                    "description": t["description"],
                    "parameters": t["input_schema"],
                },
            }
            for t in tools
        ]

    def to_anthropic_tools(self, query: str | None = None) -> list[dict[str, Any]]:
        """Format AgentGate tools for Anthropic's 'tools' parameter."""
        tools = self.discover(query)["tools"] if query else self.list_tools()["tools"]
        return [
            {
                "name": t["id"].replace(".", "__").replace("/", "---"),
                "description": t["description"],
                "input_schema": t["input_schema"],
            }
            for t in tools
        ]

    def resolve_tool_id(self, llm_tool_name: str) -> str:
        """Convert LLM-safe tool names back to AgentGate IDs."""
        return llm_tool_name.replace("---", "/").replace("__", ".")

    def discover(self, query: str, limit: int = 10) -> Dict[str, Any]:
        """Search for tools by natural-language capability."""
        response = self.client.get(
            f"{self.base_url}/discover",
            params={"q": query, "limit": limit},
            headers=self._headers(),
        )
        self._raise_for_error(response)
        return response.json()

    def connect(
        self,
        saas: str,
        owner_email: str = "owner@example.com",
        name: str | None = None,
        approval_slack_webhook_url: str | None = None,
        approval_email: str | None = None,
    ) -> Dict[str, Any]:
        """Connect a launch SaaS and get its hosted MCP gateway URL."""
        payload: Dict[str, Any] = {"owner_email": owner_email}
        if name:
            payload["name"] = name
        if approval_slack_webhook_url:
            payload["approval_slack_webhook_url"] = approval_slack_webhook_url
        if approval_email:
            payload["approval_email"] = approval_email
        response = self.client.post(
            f"{self.base_url}/connect/{saas}",
            json=payload,
            headers=self._headers(),
        )
        self._raise_for_error(response)
        return response.json()

    def approval_dashboard_url(self, api_key_in_query: bool = False) -> str:
        """Return the simple human approval dashboard URL."""
        suffix = f"?api_key={self.api_key}" if api_key_in_query else ""
        return f"{self.base_url}/dashboard/approvals{suffix}"

    def list_approvals(self) -> Dict[str, Any]:
        """List approvals and their current state."""
        response = self.client.get(f"{self.base_url}/approvals", headers=self._headers())
        self._raise_for_error(response)
        return response.json()

    def get_approval(self, approval_id: str) -> Dict[str, Any]:
        """Fetch one approval request."""
        response = self.client.get(f"{self.base_url}/approvals/{approval_id}", headers=self._headers())
        self._raise_for_error(response)
        return response.json()

    def approve(
        self,
        approval_id: str,
        reviewed_by: str = "human",
        note: str = "",
        execution_state: Dict[str, Any] | None = None,
        state: Dict[str, Any] | None = None,
    ) -> Dict[str, Any]:
        """Approve a pending action, revalidate policy, and execute if still allowed."""
        payload: Dict[str, Any] = {"reviewed_by": reviewed_by, "note": note}
        if execution_state is not None:
            payload["execution_state"] = execution_state
        elif state is not None:
            payload["state"] = state
        response = self.client.post(
            f"{self.base_url}/approvals/{approval_id}/approve",
            json=payload,
            headers=self._headers(),
        )
        self._raise_for_error(response)
        return response.json()

    def reject(self, approval_id: str, reviewed_by: str = "human", note: str = "") -> Dict[str, Any]:
        """Reject a pending action."""
        response = self.client.post(
            f"{self.base_url}/approvals/{approval_id}/reject",
            json={"reviewed_by": reviewed_by, "note": note},
            headers=self._headers(),
        )
        self._raise_for_error(response)
        return response.json()

    def traces(self, limit: int = 100) -> Dict[str, Any]:
        """List recent tool-call traces."""
        response = self.client.get(
            f"{self.base_url}/traces",
            params={"limit": limit},
            headers=self._headers(),
        )
        self._raise_for_error(response)
        return response.json()

    def export_traces(self, format: str = "json", limit: int = 100) -> Any:
        """Export traces as json, jsonl, langsmith, or helicone-compatible records."""
        response = self.client.get(
            f"{self.base_url}/traces/export",
            params={"format": format, "limit": limit},
            headers=self._headers(),
        )
        self._raise_for_error(response)
        if format == "jsonl":
            return response.text
        return response.json()

    def verify_state(
        self,
        intent: str,
        required_fields: list[str],
        assumed_state: Dict[str, Any] | None = None,
        current_state: Dict[str, Any] | None = None,
        state_refetch: Dict[str, Any] | None = None,
        conditions: Dict[str, Any] | None = None,
        on_mismatch: str = "abort",
    ) -> Dict[str, Any]:
        """Re-check critical state before execution."""
        payload: Dict[str, Any] = {
            "intent": intent,
            "required_fields": required_fields,
            "on_mismatch": on_mismatch,
        }
        if assumed_state is not None:
            payload["assumed_state"] = assumed_state
        if current_state is not None:
            payload["current_state"] = current_state
        if state_refetch is not None:
            payload["state_refetch"] = state_refetch
        if conditions is not None:
            payload["conditions"] = conditions
        response = self.client.post(f"{self.base_url}/state/verify", json=payload, headers=self._headers())
        self._raise_for_error(response)
        return response.json()

    def reconcile(self, action: Dict[str, Any] | str, **kwargs: Any) -> Dict[str, Any]:
        """Reconcile an unknown outcome before retrying an action."""
        payload: Dict[str, Any] = {"action": action}
        payload.update(kwargs)
        response = self.client.post(f"{self.base_url}/outcomes/reconcile", json=payload, headers=self._headers())
        self._raise_for_error(response)
        return response.json()

    async def acall(
        self,
        tool: str,
        params: Dict[str, Any],
        agent_id: str = "default_agent",
        policy: Dict[str, Any] | None = None,
        execution_state: Dict[str, Any] | None = None,
        action: str | None = None,
        failure_policy: Dict[str, Any] | None = None,
        entity_resolution: Dict[str, Any] | None = None,
        context: Dict[str, Any] | None = None,
        idempotency_key: str | None = None,
    ) -> Dict[str, Any]:
        """Async version."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/call",
                    json=self._payload(
                        tool,
                        params,
                        agent_id,
                        policy,
                        execution_state,
                        action,
                        failure_policy,
                        entity_resolution,
                        context,
                        idempotency_key,
                    ),
                    headers=self._headers(),
                )
            except httpx.ConnectError:
                raise AgentGateError(f"Connection refused. Is AgentGate running at {self.base_url}?")
            self._raise_for_error(response)
            return response.json()

    async def alist_tools(self) -> Dict[str, Any]:
        """Async version of list_tools."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.get(f"{self.base_url}/tools", headers=self._headers())
            self._raise_for_error(response)
            return response.json()

    async def adiscover(self, query: str, limit: int = 10) -> Dict[str, Any]:
        """Async version of discover."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.get(
                f"{self.base_url}/discover",
                params={"q": query, "limit": limit},
                headers=self._headers(),
            )
            self._raise_for_error(response)
            return response.json()

    async def aconnect(
        self,
        saas: str,
        owner_email: str = "owner@example.com",
        name: str | None = None,
        approval_slack_webhook_url: str | None = None,
        approval_email: str | None = None,
    ) -> Dict[str, Any]:
        """Async version of connect."""
        payload: Dict[str, Any] = {"owner_email": owner_email}
        if name:
            payload["name"] = name
        if approval_slack_webhook_url:
            payload["approval_slack_webhook_url"] = approval_slack_webhook_url
        if approval_email:
            payload["approval_email"] = approval_email
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                f"{self.base_url}/connect/{saas}",
                json=payload,
                headers=self._headers(),
            )
            self._raise_for_error(response)
            return response.json()

    async def alist_approvals(self) -> Dict[str, Any]:
        """Async version of list_approvals."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.get(f"{self.base_url}/approvals", headers=self._headers())
            self._raise_for_error(response)
            return response.json()

    async def aapprove(
        self,
        approval_id: str,
        reviewed_by: str = "human",
        note: str = "",
        execution_state: Dict[str, Any] | None = None,
        state: Dict[str, Any] | None = None,
    ) -> Dict[str, Any]:
        """Async version of approve."""
        payload: Dict[str, Any] = {"reviewed_by": reviewed_by, "note": note}
        if execution_state is not None:
            payload["execution_state"] = execution_state
        elif state is not None:
            payload["state"] = state
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                f"{self.base_url}/approvals/{approval_id}/approve",
                json=payload,
                headers=self._headers(),
            )
            self._raise_for_error(response)
            return response.json()

    async def areject(self, approval_id: str, reviewed_by: str = "human", note: str = "") -> Dict[str, Any]:
        """Async version of reject."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.post(
                f"{self.base_url}/approvals/{approval_id}/reject",
                json={"reviewed_by": reviewed_by, "note": note},
                headers=self._headers(),
            )
            self._raise_for_error(response)
            return response.json()

    async def atraces(self, limit: int = 100) -> Dict[str, Any]:
        """Async version of traces."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.get(
                f"{self.base_url}/traces",
                params={"limit": limit},
                headers=self._headers(),
            )
            self._raise_for_error(response)
            return response.json()

    async def aexport_traces(self, format: str = "json", limit: int = 100) -> Any:
        """Async version of export_traces."""
        async with httpx.AsyncClient(timeout=120.0) as client:
            response = await client.get(
                f"{self.base_url}/traces/export",
                params={"format": format, "limit": limit},
                headers=self._headers(),
            )
            self._raise_for_error(response)
            if format == "jsonl":
                return response.text
            return response.json()

    def close(self) -> None:
        self.client.close()


# Convenience instances. Prefer `trace`; provide `gate` as a compatibility alias.
trace = Trace.from_env()
gate = trace

File: src/sdk.ts
export type JsonObject = Record<string, unknown>;

export type AgentGateOptions = {
  apiKey: string;
  baseUrl?: string;
};

export type ExecutionContext = {
  env?: string;
  workflowId?: string;
  workflowRunId?: string;
  userId?: string;
  workspaceId?: string;
  resources?: JsonObject;
  labels?: JsonObject;
  traceId?: string;
};

export type CallOptions = {
  tool: string;
  params: JsonObject;
  agentId?: string;
  policy?: JsonObject;
  executionState?: JsonObject;
  action?: string;
  failurePolicy?: JsonObject;
  entityResolution?: JsonObject;
  context?: ExecutionContext;
  idempotencyKey?: string;
};

export type ConnectOptions = {
  saas: string;
  ownerEmail?: string;
  name?: string;
  approvalSlackWebhookUrl?: string;
  approvalEmail?: string;
};

export type VerifyStateOptions = {
  intent: string;
  requiredFields: string[];
  assumedState?: JsonObject;
  currentState?: JsonObject;
  stateRefetch?: JsonObject;
  conditions?: JsonObject;
  threshold?: string;
  onMismatch?: string;
  params?: JsonObject;
  agentId?: string;
};

export class AgentGate {
  private readonly apiKey: string;
  private readonly baseUrl: string;

  constructor(options: AgentGateOptions) {
    if (!options.apiKey || !options.apiKey.trim()) {
      throw new Error("AgentGate requires a non-empty API key");
    }
    this.apiKey = options.apiKey.trim();
    this.baseUrl = (options.baseUrl ?? "http://localhost:8000").replace(/\/+$/, "");
  }

  static fromEnv(): AgentGate {
    const apiKey = process.env.TRACE_API_KEY ?? process.env.AGENTGATE_API_KEY;
    const baseUrl = process.env.TRACE_BASE_URL ?? process.env.AGENTGATE_BASE_URL;
    if (!apiKey) {
      throw new Error("Set TRACE_API_KEY or AGENTGATE_API_KEY, or pass { apiKey } to AgentGate");
    }
    return new AgentGate({ apiKey, baseUrl });
  }

  approvalDashboardUrl(apiKeyInQuery = false): string {
    return `${this.baseUrl}/dashboard/approvals${apiKeyInQuery ? `?api_key=${encodeURIComponent(this.apiKey)}` : ""}`;
  }

  private headers(): HeadersInit {
    return {
      "Content-Type": "application/json",
      "X-API-Key": this.apiKey,
    };
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        ...this.headers(),
        ...(init.headers ?? {}),
      },
    });
    if (!response.ok) {
      let detail: unknown;
      try {
        detail = await response.json();
      } catch {
        detail = await response.text();
      }
      throw new Error(`AgentGate error ${response.status}: ${JSON.stringify(detail)}`);
    }
    return response.json() as Promise<T>;
  }

  call(options: CallOptions): Promise<JsonObject> {
    return this.request<JsonObject>("/call", {
      method: "POST",
      body: JSON.stringify({
        tool: options.tool,
        params: options.params,
        agent_id: options.agentId ?? "default_agent",
        ...(options.policy ? { policy: options.policy } : {}),
        ...(options.executionState ? { execution_state: options.executionState } : {}),
        ...(options.action ? { action: options.action } : {}),
        ...(options.failurePolicy ? { failure_policy: options.failurePolicy } : {}),
        ...(options.entityResolution ? { entity_resolution: options.entityResolution } : {}),
        ...(options.context ? { context: normalizeContext(options.context) } : {}),
        ...(options.idempotencyKey ? { idempotency_key: options.idempotencyKey } : {}),
      }),
    });
  }

  listTools(): Promise<JsonObject> {
    return this.request<JsonObject>("/tools");
  }

  discover(query: string, limit = 10): Promise<JsonObject> {
    const params = new URLSearchParams({ q: query, limit: String(limit) });
    return this.request<JsonObject>(`/discover?${params.toString()}`);
  }

  connect(options: ConnectOptions): Promise<JsonObject> {
    return this.request<JsonObject>(`/connect/${encodeURIComponent(options.saas)}`, {
      method: "POST",
      body: JSON.stringify({
        owner_email: options.ownerEmail ?? "owner@example.com",
        ...(options.name ? { name: options.name } : {}),
        ...(options.approvalSlackWebhookUrl ? { approval_slack_webhook_url: options.approvalSlackWebhookUrl } : {}),
        ...(options.approvalEmail ? { approval_email: options.approvalEmail } : {}),
      }),
    });
  }

  listApprovals(): Promise<JsonObject> {
    return this.request<JsonObject>("/approvals");
  }

  getApproval(approvalId: string): Promise<JsonObject> {
    return this.request<JsonObject>(`/approvals/${encodeURIComponent(approvalId)}`);
  }

  approve(
    approvalId: string,
    body: { reviewedBy?: string; note?: string; executionState?: JsonObject; state?: JsonObject } = {},
  ): Promise<JsonObject> {
    return this.request<JsonObject>(`/approvals/${encodeURIComponent(approvalId)}/approve`, {
      method: "POST",
      body: JSON.stringify({
        reviewed_by: body.reviewedBy ?? "human",
        note: body.note ?? "",
        ...(body.executionState ? { execution_state: body.executionState } : {}),
        ...(body.state ? { state: body.state } : {}),
      }),
    });
  }

  reject(approvalId: string, reviewedBy = "human", note = ""): Promise<JsonObject> {
    return this.request<JsonObject>(`/approvals/${encodeURIComponent(approvalId)}/reject`, {
      method: "POST",
      body: JSON.stringify({ reviewed_by: reviewedBy, note }),
    });
  }

  verifyState(options: VerifyStateOptions): Promise<JsonObject> {
    return this.request<JsonObject>("/state/verify", {
      method: "POST",
      body: JSON.stringify({
        intent: options.intent,
        required_fields: options.requiredFields,
        threshold: options.threshold ?? "strict",
        on_mismatch: options.onMismatch ?? "abort",
        agent_id: options.agentId ?? "default_agent",
        ...(options.assumedState ? { assumed_state: options.assumedState } : {}),
        ...(options.currentState ? { current_state: options.currentState } : {}),
        ...(options.stateRefetch ? { state_refetch: options.stateRefetch } : {}),
        ...(options.conditions ? { conditions: options.conditions } : {}),
        ...(options.params ? { params: options.params } : {}),
      }),
    });
  }

  reconcile(action: JsonObject | string): Promise<JsonObject> {
    return this.request<JsonObject>("/outcomes/reconcile", {
      method: "POST",
      body: JSON.stringify(typeof action === "string" ? { action } : action),
    });
  }

  traces(limit = 100): Promise<JsonObject> {
    return this.request<JsonObject>(`/traces?${new URLSearchParams({ limit: String(limit) }).toString()}`);
  }

  async exportTraces(format = "json", limit = 100): Promise<JsonObject | string> {
    const params = new URLSearchParams({ format, limit: String(limit) });
    const response = await fetch(`${this.baseUrl}/traces/export?${params.toString()}`, {
      headers: this.headers(),
    });
    if (!response.ok) {
      throw new Error(`AgentGate error ${response.status}: ${await response.text()}`);
    }
    return format === "jsonl" ? response.text() : response.json();
  }
}

function normalizeContext(context: ExecutionContext): JsonObject {
  return {
    ...(context.env ? { env: context.env } : {}),
    ...(context.workflowId ? { workflow_id: context.workflowId } : {}),
    ...(context.workflowRunId ? { workflow_run_id: context.workflowRunId } : {}),
    ...(context.userId ? { user_id: context.userId } : {}),
    ...(context.workspaceId ? { workspace_id: context.workspaceId } : {}),
    ...(context.resources ? { resources: context.resources } : {}),
    ...(context.labels ? { labels: context.labels } : {}),
    ...(context.traceId ? { trace_id: context.traceId } : {}),
  };
}
